File: 2015/Haskell/src/Day23.hs
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE TemplateHaskell #-}

module Day23
  ( day23,
  )
where

import AoC
import AoC.Common (Parser, unwrapParser)
import Control.Applicative ((<|>))
import Control.Lens
import Data.Functor (($>))
import Data.Text (Text)
import qualified Data.Text as Text
import Data.Vector (Vector)
import qualified Data.Vector as Vector
import Text.Megaparsec (parse)
import Text.Megaparsec.Char (char, space1, string)
import qualified Text.Megaparsec.Char.Lexer as L

type Parsed = Vector Instruction

type Offset = Int

data Register = A | B deriving (Show)

data Instruction = Hlf Register | Tpl Register | Inc Register | Jmp Offset | Jie Register Offset | Jio Register Offset deriving (Show)

pInstr :: Parser Instruction
pInstr =
  do
    constructor <- pComplexJmp
    _ <- space1
    reg <- pRegister
    _ <- char ','
    _ <- space1
    constructor reg <$> pOffset
    <|> do
      constructor <- pJmp
      _ <- space1
      constructor <$> pOffset
    <|> do
      constructor <- pRegisterOpcode
      _ <- space1
      constructor <$> pRegister

pRegisterOpcode :: Parser (Register -> Instruction)
pRegisterOpcode =
  string "hlf" $> Hlf
    <|> string "tpl" $> Tpl
    <|> string "inc" $> Inc

pJmp :: Parser (Offset -> Instruction)
pJmp = string "jmp" $> Jmp

pComplexJmp :: Parser (Register -> Offset -> Instruction)
pComplexJmp = string "jie" $> Jie <|> string "jio" $> Jio

pRegister :: Parser Register
pRegister =
  char 'a' $> A
    <|> char 'b' $> B

pOffset :: Parser Offset
pOffset = do
  sign <- pSign
  (* sign) <$> L.decimal

pSign :: Parser Int
pSign =
  char '+' $> 1
    <|> char '-' $> (-1)

data VM = VM
  { _a :: Int,
    _b :: Int,
    _pc :: Int
  }
  deriving (Show)

makeLenses ''VM

f :: VM -> Instruction -> VM
f vm instruction = case instruction of
  Hlf r -> vm & pc %~ (+ 1) & r %%~ (`div` 2)
  Tpl r -> vm & pc %~ (+ 1) & r %%~ (* 3)
  Inc r -> vm & pc %~ (+ 1) & r %%~ (+ 1)
  Jmp off -> vm & (pc %~ (+ off))
  Jie r off ->
    if even $ vm ^^. r
      then vm & pc %~ (+ off)
      else vm & pc %~ (+ 1)
  Jio r off ->
    if vm ^^. r == 1
      then vm & pc %~ (+ off)
      else vm & pc %~ (+ 1)
  where
    reg A = a
    reg B = b
    a ^^. b = a ^. reg b
    a %%~ b = reg a %~ b

runVM :: VM -> Parsed -> VM
runVM vm instructions = case instructions Vector.!? (vm ^. pc) of
  Nothing -> vm
  Just inst -> runVM (f vm inst) instructions

day23 :: Runner Parsed Int
day23 =
  let year = 2015
      day = 23
      parser :: Text -> Maybe Parsed
      parser = unwrapParser . traverse (parse pInstr "(input)") . Vector.fromList . Text.lines
      part1 :: Parsed -> Maybe Int
      part1 = return . (^. b) . runVM (VM 0 0 0)
      part2 :: Parsed -> Maybe Int
      part2 = return . (^. b) . runVM (VM 1 0 0)
   in Runner {..}

File: 2015/Rust/src/day23.rs
use std::ops::{Index, IndexMut};

use crate::Result;

#[derive(Debug, Clone, Copy)]
#[repr(usize)]
pub enum Register {
    A = 0,
    B = 1,
}

impl From<&str> for Register {
    fn from(value: &str) -> Self {
        match value {
            "a" => Self::A,
            "b" => Self::B,
            _ => panic!(),
        }
    }
}

#[derive(Debug, Clone, Copy)]
pub enum Instruction {
    Hlf(Register),
    Tpl(Register),
    Inc(Register),
    Jmp(isize),
    Jie(Register, isize),
    Jio(Register, isize),
}

pub fn parse(inpt: &str) -> Vec<Instruction> {
    fn p_offset(off: &str) -> isize {
        off[1..].parse::<isize>().unwrap()
            * match off.as_bytes()[0] {
                b'+' => 1,
                b'-' => -1,
                _ => panic!(),
            }
    }
    inpt.lines()
        .map(|line| match &line[..3] {
            "hlf" => Instruction::Hlf(line[4..5].into()),
            "tpl" => Instruction::Tpl(line[4..5].into()),
            "inc" => Instruction::Inc(line[4..5].into()),
            "jmp" => Instruction::Jmp(p_offset(&line[4..])),
            "jie" => Instruction::Jie(line[4..5].into(), p_offset(&line[7..])),
            "jio" => Instruction::Jio(line[4..5].into(), p_offset(&line[7..])),
            _ => panic!(),
        })
        .collect()
}

impl Index<Register> for [usize; 2] {
    type Output = usize;

    fn index(&self, index: Register) -> &Self::Output {
        &self[index as usize]
    }
}

impl IndexMut<Register> for [usize; 2] {
    fn index_mut(&mut self, index: Register) -> &mut Self::Output {
        &mut self[index as usize]
    }
}

struct VM {
    a: usize,
    b: usize,
    pc: usize,
}
impl Index<Register> for VM {
    type Output = usize;

    fn index(&self, index: Register) -> &Self::Output {
        match index {
            Register::A => &self.a,
            Register::B => &self.b,
        }
    }
}

impl IndexMut<Register> for VM {
    fn index_mut(&mut self, index: Register) -> &mut Self::Output {
        match index {
            Register::A => &mut self.a,
            Register::B => &mut self.b,
        }
    }
}

fn run_vm(instructions: &[Instruction], mut state: VM) -> VM {
    while let Some(&instruction) = instructions.get(state.pc) {
        match instruction {
            Instruction::Hlf(r) => {
                state[r] /= 2;
                state.pc += 1;
            }
            Instruction::Tpl(r) => {
                state[r] *= 3;
                state.pc += 1;
            }
            Instruction::Inc(r) => {
                state[r] += 1;
                state.pc += 1;
            }
            Instruction::Jmp(off) => state.pc = state.pc.wrapping_add_signed(off),
            Instruction::Jie(r, off) => {
                state.pc = state
                    .pc
                    .wrapping_add_signed(if state[r] % 2 == 0 { off } else { 1 })
            }
            Instruction::Jio(r, off) => {
                state.pc = state
                    .pc
                    .wrapping_add_signed(if state[r] == 1 { off } else { 1 })
            }
        }
    }

    state
}

pub fn part1(parsed: &[Instruction]) -> usize {
    run_vm(parsed, VM { a: 0, b: 0, pc: 0 }).b
}

pub fn part2(parsed: &[Instruction]) -> usize {
    run_vm(parsed, VM { a: 1, b: 0, pc: 0 }).b
}

pub fn run(input: &str) -> Result<()> {
    let parsed = parse(input);
    println!("Part 1: {}", part1(&parsed));
    println!("Part 2: {}", part2(&parsed));

    Ok(())
}
